Suppress stdout from cython layer

sergeant-wizard opened the pull request:
For #11729 


-   [X] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [X] Title summarizes what is changing.
-   [x] Has a [news entry](https://github.com/Microsoft/vscode-python/tree/master/news) file (remember to thank yourself!).
-   [X] Appropriate comments and documentation strings in the code.
-   [X] ~Has sufficient logging.~
-   [x] Has telemetry for enhancements.
-   [X] Unit tests & system/integration tests are added/updated.
-   [x] ~[Test plan](https://github.com/Microsoft/vscode-python/blob/master/.github/test_plan.md) is updated as appropriate.~
-   [x] ~[`package-lock.json`](https://github.com/Microsoft/vscode-python/blob/master/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).~
-   [x] ~The wiki is updated with any design decisions/details.~


I'm using pybullet for a project and it prints to stdout by simply running `import pybullet`.
What's more it can't be suppressed by the current `utils.hide_stdio` method because the message is coming from C code via Cython.
This patch will suppress messages from Cython layer by swapping the lower level file descriptors (I learned about the idea from [here](https://stackoverflow.com/questions/5081657/how-do-i-prevent-a-c-shared-library-to-print-on-stdout-in-python/17954769#17954769).
It's a dirty workaround but did work for my particular use case.
That said I think #6594 is probably the better approach.

I wasn't sure if I should send this PR as something to be considered merging, or if I should just inform users encountering #11729 to apply this patch as a temporary workaround.
I appreciate it if you could tell me what you think.

int19h commented on pythonFiles/testing_tools/adapter/util.py:
```diff
@@ -173,14 +173,19 @@ def fix_fileid(
+        finally:
+            os.dup2(old_stdout.fileno(), stdout_fd)
+            sys.stdout = os.fdopen(stdout_fd, "w")
+            sys.stderr = sys.__stderr__
```
The `finally` part won't run at all if either the first `fdopen` or the first `dup` call fail. Given that `sys.stdout` and `sys.stderr` have already been assigned, it ought to at least reverse those.

I think this needs two separate `try..finally` blocks here - one that goes around `with`, and that restores `sys` members to their old values, and the other one inside `with` to restore the stdout file descriptor. I don't think it needs to re-open the descriptor to create a new `sys.stdout` stream (indeed, I don't think it'll accurately re-create it on all platforms) - since the actual descriptor value remains the same, it should be possible to just stash away the old stdout/stderr objects, and restore them.

int19h commented on pythonFiles/testing_tools/adapter/util.py:
```diff
@@ -173,14 +173,19 @@ def fix_fileid(
 @contextlib.contextmanager
 def hide_stdio():
     """Swallow stdout and stderr."""
```
Given the description, I think this needs to do the same thing for stdout and stderr - i.e. apply the low-level fd magic to the latter, as well.

sergeant-wizard commented:
Thanks for the feedback.
I made the changes you suggested.
I guess it's worth updating the tests so this can be merged?

sergeant-wizard commented:
I'm not sure if the "Windows Py3x Single Workspace" test is failing because of the changes.
I checked on my windows environment by directly editting the installed `util.py` that test discover succeeds.

ericsnowcurrently commented on pythonFiles/testing_tools/adapter/util.py:
```diff
@@ -170,17 +170,36 @@ def fix_fileid(
+def _hide_buffer(attr, ignored):
+    # an alternative approach is suggested: #6594
+    assert attr in ("stdout", "stderr")
+    stdout_fd = getattr(sys, attr).fileno()
```
`sys.stdout` and `sys.stderr` may have already been set to something that does not have `fileno()`, so this line out fail.  In that situation you'd want to either not worry about swapping the FDs or use `sys.__stdout__.fileno()`, etc. (or just hard-code it, e.g. `1` for stdout and `2` for stderr).

ericsnowcurrently commented on pythonFiles/tests/testing_tools/adapter/pytest/test_discovery.py:
```diff
@@ -335,14 +337,23 @@ def fake_pytest_main(args, plugins):
             calls.insert(3, ("discovered.__len__", None, None))
 
-        buf = StringIO()
+        # to simulate stdio behavior including methods like .fileno(),
+        # use actual files (rather than StringIO)
```
I recommend a separate test for each case.  The existing code verifies the file behavior.  The new test would verify the file descriptor behavior.

sergeant-wizard replied:
I split the helpers like you suggested (thanks!).
Are you suggesting that there should now be unit tests for `_replace_fd` and `_replace_stdx` methods?
The tests found here seem to be for outward facing public methods like `hide_stdio` but I'm happy to add tests for the two helper methods as well.

ericsnowcurrently commented on pythonFiles/testing_tools/adapter/util.py:
```diff
@@ -170,17 +170,36 @@ def fix_fileid(
 @contextlib.contextmanager
 def hide_stdio():
     """Swallow stdout and stderr."""
     ignored = StdioStream()
```
Rather than `StdioStream` we could just use the devnull file object you create in `_hide_buffer()`.  The down-side is that it would add a new FS-related possible failure point.  However, the up-side is that we could get rid of `StdioStream` and the related import shenanigans we do at the top of the file.

sergeant-wizard replied:
The contents of the stream created here will be printed upon an error [here](https://github.com/microsoft/vscode-python/blob/fec21d0288bcb1665f766006f77a3fd03eec3095/pythonFiles/testing_tools/adapter/pytest/_discovery.py#L42). Can we still do this if we replace this with devnull opened in write mode?

sergeant-wizard replied:
I realized this could work if I use temporary files instead of `os.devnull`, so I modified the code accordingly.

ericsnowcurrently commented on pythonFiles/testing_tools/adapter/util.py:
```diff
@@ -170,17 +170,36 @@ def fix_fileid(
+    # This will swallow python print statements, logs, etc.
+    setattr(sys, attr, ignored)
+    try:
+        with os.fdopen(os.dup(stdout_fd), "w") as old_stdout:
+            # Set / unset file descriptors.
+            # This is required for output coming out of layers lower than python,
+            # e.g., C printf statements via cython.
+            with open(os.devnull, "w") as devnull:
+                os.dup2(devnull.fileno(), stdout_fd)
+            try:
+                yield ignored
+            finally:
+                os.dup2(old_stdout.fileno(), stdout_fd)
```
I was going to say something about making sure the dup FD gets closed (so we don't leak it).  However, `with os.fdopen():` takes care of that.  Then again, there is no need to use `os.fdopen()`.  (It's also good to do only one thing per line.)

```suggestion
        dup_fd = os.dup(stdout_fd)
        try:
            # Set / unset file descriptors.
            # This is required for output coming out of layers lower than python,
            # e.g., C printf statements via cython.
            with open(os.devnull, "w") as devnull:
                os.dup2(devnull.fileno(), stdout_fd)
            try:
                yield ignored
            finally:
                os.dup2(dup_fd, stdout_fd)
        finally:
            os.close(dup_fd)
```

ericsnowcurrently replied:
As it is written, this can confuse readers that don't know how file descriptors work.  So let's make it more clear:

```suggestion
            with open(os.devnull, "w") as devnull:
                os.dup2(devnull.fileno(), stdout_fd)
                try:
                    yield ignored
                finally:
                    os.dup2(old_stdout.fileno(), stdout_fd)
```

sergeant-wizard commented:
Thanks for your comments and suggestions.

* I split helper context managers into smaller pieces like you suggested (thanks, the code is easier to read)
* I removed the `StringIO` all together and switched to temporary files to swallow the buffer
* Instead of using `os.devnull` as the alternative file descriptor, I pointed it to the temporary file above so I can retrieve the contents
* I added test cases that would simulate the behavior where file descriptor swapping is needed

ericsnowcurrently commented on pythonFiles/testing_tools/adapter/util.py:
```diff
@@ -171,27 +168,79 @@ def fix_fileid(
 
+class _StringIO:
+    def setvalue(self, value):
+        self._value = value
 
-else:
-    StdioStream = StringIO
+    def getvalue(self):
+        return self._value
```
Why not use `StringIO` directly?  (leave the original import statement)

ericsnowcurrently commented on pythonFiles/testing_tools/adapter/util.py:
```diff
@@ -171,27 +168,79 @@ def fix_fileid(
+        return self._value
+
+
+@contextlib.contextmanager
+def _temp_io():
+    td = tempfile.mkdtemp()
+    path = os.path.join(td, "temp")
+    sio = _StringIO()
+    try:
+        with open(path, "w") as f:
+            yield sio, f
+    finally:
+        with open(path, "r") as f:
+            sio.setvalue(f.read())
+        try:
+            os.remove(path)
+        except OSError:
+            pass
+        os.rmdir(td)
```
You could also use `tempfile.TemporaryFile`:

```suggestion
@contextlib.contextmanager
def _temp_io():
    sio = _StringIO()
    with tempfile.TemporaryFile() as tmp:
        try:
            yield sio, tmp
        finally:
            tmp.seek(0)
            sio.setvalue(tmp.read())
```

sergeant-wizard replied:
Changed to "r+" mode.

ericsnowcurrently commented on pythonFiles/tests/testing_tools/adapter/pytest/test_discovery.py:
```diff
@@ -251,6 +253,19 @@ class DiscoverTests(unittest.TestCase):
     ]
+    pytest_stdout = "spamspamspamspamspamspamspammityspam"
+
+    @staticmethod
```
`staticmethod` is rarely the right solution.  I recommend making this a module-level function.

ericsnowcurrently commented on pythonFiles/tests/testing_tools/adapter/pytest/test_discovery.py:
```diff
@@ -314,14 +329,8 @@ def test_no_tests_found(self):
         self.assertEqual(stub.calls, calls)
 
     def test_stdio_hidden(self):
-        pytest_stdout = "spamspamspamspamspamspamspammityspam"
```
I recommend leaving test data as close to (or in) the test as possible.  So leave this here.

ericsnowcurrently commented on pythonFiles/tests/testing_tools/adapter/pytest/test_discovery.py:
```diff
@@ -335,27 +344,45 @@ def fake_pytest_main(args, plugins):
         self.assertEqual(stub.calls, calls)
 
     def test_stdio_not_hidden(self):
-        pytest_stdout = "spamspamspamspamspamspamspammityspam"
```
I recommend leaving test data as close to (or in) the test as possible.  So leave this here.

ericsnowcurrently commented on pythonFiles/tests/testing_tools/adapter/pytest/test_discovery.py:
```diff
@@ -251,6 +253,19 @@ class DiscoverTests(unittest.TestCase):
+    pytest_stdout = "spamspamspamspamspamspamspammityspam"
+
+    @staticmethod
+    def fake_pytest_main(stub, use_fd):
```
```suggestion
    def fake_pytest_main(stub, use_fd, pytest_stdout):
```

ericsnowcurrently commented on pythonFiles/tests/testing_tools/adapter/pytest/test_discovery.py:
```diff
@@ -335,27 +344,45 @@ def fake_pytest_main(args, plugins):
+        discover(
+            [],
+            hidestdio=True,
+            _pytest_main=self.fake_pytest_main(stub, True),
```
```suggestion
            _pytest_main=self.fake_pytest_main(stub, True, pytest_stdout),
```

ericsnowcurrently commented on pythonFiles/tests/testing_tools/adapter/pytest/test_discovery.py:
```diff
@@ -373,12 +400,29 @@ def fake_pytest_main(args, plugins):
+            discover(
+                [],
+                hidestdio=False,
+                _pytest_main=self.fake_pytest_main(stub, False),
```
```suggestion
                _pytest_main=self.fake_pytest_main(stub, False, pytest_stdout),
```

ericsnowcurrently commented on pythonFiles/tests/testing_tools/adapter/pytest/test_discovery.py:
```diff
@@ -335,27 +344,45 @@ def fake_pytest_main(args, plugins):
-        buf = StringIO()
+        # to simulate stdio behavior including methods like .fileno(),
+        # use actual files (rather than StringIO)
+        td = tempfile.mkdtemp()
+        stdio_mock = path.join(td, "stdio_mock")
+        with open(stdio_mock, "w") as mock:
```
Again, `tempfile.TemporaryFile` should simplify things for you.

ericsnowcurrently commented on pythonFiles/tests/testing_tools/adapter/pytest/test_discovery.py:
```diff
@@ -335,27 +344,45 @@ def fake_pytest_main(args, plugins):
+        self.assertEqual(captured, "")
+        self.assertEqual(stub.calls, calls)
 
+        # simulate cases where stdout comes from the lower layer than sys.stdout
+        # via file descriptors (e.g., from cython)
+        stub.calls = []
+        discover(
+            [],
+            hidestdio=True,
+            _pytest_main=self.fake_pytest_main(stub, True),
+            _plugin=plugin,
+        )
         self.assertEqual(captured, "")
         self.assertEqual(stub.calls, calls)
```
Put this in a separate test method.

ericsnowcurrently commented on pythonFiles/tests/testing_tools/adapter/pytest/test_discovery.py:
```diff
@@ -373,12 +400,29 @@ def fake_pytest_main(args, plugins):
+        self.assertEqual(captured, DiscoverTests.pytest_stdout)
+        self.assertEqual(stub.calls, calls)
+
+        # simulate cases where stdout comes from the lower layer than sys.stdout
+        # via file descriptors (e.g., from cython)
+        stub.calls = []
+        discover(
+            [],
+            hidestdio=False,
+            _pytest_main=self.fake_pytest_main(stub, True),
+            _plugin=plugin,
+        )
+        self.assertEqual(captured, DiscoverTests.pytest_stdout)
         self.assertEqual(stub.calls, calls)
```
Put this in a separate test method.

sergeant-wizard commented on pythonFiles/testing_tools/adapter/util.py:
```diff
@@ -171,27 +168,79 @@ def fix_fileid(
 
-if sys.version_info < (3,):
+@contextlib.contextmanager
+def _replace_stdx(attr, target):
```
OK I made two separate functions.

sergeant-wizard commented on pythonFiles/testing_tools/adapter/util.py:
```diff
@@ -171,27 +168,79 @@ def fix_fileid(
+    target_fd = target.fileno()
+
+    # `os.dup2()` closes the original FD, so we make copies.
+    dup_fd = os.dup(fd)
```
I updated the comment, was that your intention?
